//---------------------------------------------------------------------
// Copyright 2020 National Instruments
//
// Licensed under the MIT License;
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     https://mit-license.org/
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//---------------------------------------------------------------------

//---------------------------------------------------------------------
//---------------------------------------------------------------------
syntax = "proto3";

//---------------------------------------------------------------------
//---------------------------------------------------------------------
option java_multiple_files = true;
option java_package = "labview.measurementservice";
option java_outer_classname = "MeasurementService";
option objc_class_prefix = "LVMS";

//---------------------------------------------------------------------
//---------------------------------------------------------------------
package measurementservice;

//---------------------------------------------------------------------
// The MeasurementService service definition.
//---------------------------------------------------------------------
service MeasurementService {
  
  rpc Invoke (InvokeRequest) returns (InvokeResponse) {}
  
  rpc Query (QueryRequest) returns (QueryResponse) {}
  
  rpc Register (RegistrationRequest) returns (stream ServerEvent) {}
  
  rpc OpenDMMSession (DMMOpenSessionRequest) returns (DMMOpenSessionResponse) {}
  
  rpc ConfigureDMM (DMMConfigRequest) returns (DMMConfigResponse) {}
  
  rpc PerformDMMMeasurement (DMMMeasurementRequest) returns (DMMMeasurementData) {}
  
  rpc CloseDMMSession (DMMCloseSessionRequest) returns (DMMCloseSessionResponse) {}
  
  rpc SelfCalibrateDMM (DMMSelfCalRequest) returns (DMMSelfCalResponse) {}
  
  rpc ResetDMM (DMMResetRequest) returns (DMMResetResponse) {}
  
  rpc GetDMMLastExtCalInfo (DMMLastExtCalInfoRequest) returns (DMMLastExtCalInfoResponse) {}
  
  rpc GetDMMLastSelfCalInfo (DMMLastSelfCalInfoRequest) returns (DMMLastSelfCalInfoResponse) {}
}

//---------------------------------------------------------------------
//---------------------------------------------------------------------
message ErrorOut {
  int64 errCode = 1;
  string errMessage = 2;
}

//---------------------------------------------------------------------
//---------------------------------------------------------------------
message DMMOpenSessionRequest {
  string DMMName = 1;
}

//---------------------------------------------------------------------
//---------------------------------------------------------------------
message DMMOpenSessionResponse {
  ErrorOut error = 1;
}

//---------------------------------------------------------------------
//---------------------------------------------------------------------
message DMMConfigRequest {
  int64 DMMFunction = 1;
  double DMMResolution = 2;
  double DMMRange = 3;
  int64 DMMSampleCount = 4;
  int64 DMMApertureTimeUnit = 5;
  double DMMApertureTime = 6;
  int64 DMMNumberofAverages = 7;
  int64 DMMAutoZero = 8;
  int64 DMMADCCalibration = 9;
  double DMMSettleTime = 10;
}

//---------------------------------------------------------------------
//---------------------------------------------------------------------
message DMMConfigResponse {
  int64 DMMFunction = 1;
  double DMMResolution = 2;
  double DMMRange = 3;
  int64 DMMSampleCount = 4;
  int64 DMMApertureTimeUnit = 5;
  double DMMApertureTime = 6;
  int64 DMMNumberofAverages = 7;
  int64 DMMAutoZero = 8;
  int64 DMMADCCalibration = 9;
  double DMMSettleTime = 10;
  ErrorOut error = 11;
}

//---------------------------------------------------------------------
//---------------------------------------------------------------------
message DMMMeasurementRequest {
}

//---------------------------------------------------------------------
//---------------------------------------------------------------------
message DMMMeasurementData {
  repeated double measurement = 1;
  ErrorOut error = 2;
}

//---------------------------------------------------------------------
//---------------------------------------------------------------------
message DMMCloseSessionRequest {
}

//---------------------------------------------------------------------
//---------------------------------------------------------------------
message DMMCloseSessionResponse {
  ErrorOut error = 1;
}

//---------------------------------------------------------------------
//---------------------------------------------------------------------
message DMMSelfCalRequest {
}

//---------------------------------------------------------------------
//---------------------------------------------------------------------
message DMMSelfCalResponse {
  ErrorOut error = 1;
}

//---------------------------------------------------------------------
//---------------------------------------------------------------------
message DMMResetRequest {
}

//---------------------------------------------------------------------
//---------------------------------------------------------------------
message DMMResetResponse {
  ErrorOut error = 1;
}

//---------------------------------------------------------------------
//---------------------------------------------------------------------
message DMMLastExtCalInfoRequest {
}

//---------------------------------------------------------------------
//---------------------------------------------------------------------
message DMMLastExtCalInfoResponse {
  int64 month = 1;
  int64 day = 2;
  int64 year = 3;
  int64 hour = 4;
  int64 minute = 5;
  ErrorOut error = 6;
}

//---------------------------------------------------------------------
//---------------------------------------------------------------------
message DMMLastSelfCalInfoRequest {
}

//---------------------------------------------------------------------
//---------------------------------------------------------------------
message DMMLastSelfCalInfoResponse {
  int64 month = 1;
  int64 day = 2;
  int64 year = 3;
  int64 hour = 4;
  int64 minute = 5;
  ErrorOut error = 6;
}

//---------------------------------------------------------------------
//---------------------------------------------------------------------
message InvokeRequest {
  string command = 1;
  string parameter = 2;
}

//---------------------------------------------------------------------
//---------------------------------------------------------------------
message InvokeResponse {
  int64 status = 1;
}

//---------------------------------------------------------------------
//---------------------------------------------------------------------
message QueryRequest {
  string query = 1;
}

//---------------------------------------------------------------------
//---------------------------------------------------------------------
message QueryResponse {
  string message = 1;
  int64 status = 2;
}

//---------------------------------------------------------------------
//---------------------------------------------------------------------
message RegistrationRequest {
  string eventName = 1;
}

//---------------------------------------------------------------------
//---------------------------------------------------------------------
message ServerEvent {
  string eventData = 1;
  int64 serverId = 2;
  int64 status = 3;
}
